Add GCS signed URL support

l1nxy opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #5233.

# Rationale for this change
 
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->
Adding support for GCS Signed URL generation.
# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?


<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

l1nxy commented:
@tustvold Hi, I have completed the first version of GCP signed URL generation. Could you please take some time to review it?  and then i can update the code gradually.
Regarding the credential structure, I think we need to add the client email field. This is because the signed URL generation requires the client email, not just the sign blob method. Even when generating a signed URL with an HMAC key, we still need the client email to create credentials with scope,  the code example showed.

https://github.com/GoogleCloudPlatform/python-docs-samples/blob/6e591bec774bf0f3bf53864146f24028b98ac8df/storage/signed_urls/generate_signed_urls.py#L63-L65

next, i will add test for the sign blob and add code for using HAMC key to generating signed URL.

tustvold commented:
Sorry for the delay in reviewing this, I wonder if instead of modifying `GcpCredential` which is a breaking change, we could introduce a `GcpSigningCredential` with something like

```
struct GcpSigningCredential {
    credential: GcpCredential,
    email: String,
}
```

This could then have a separate `CredentialProvider<GcpSigningCredential>`

l1nxy commented:
I completely agree with you. However, a question arises: if I add a new credential type, will all implementations for TokenProvider need to be rewritten for GcpSigningCredential? This is because it involves additional operations to retrieve the client's email address. I am wondering if this would be acceptable.

tustvold commented:
I would think you could use composition to avoid code duplication, with the GcpSigningCredential implementation calling through to the GcpCredential implementation

s-i-l-k-e commented:
@l1nxy how is this going? Let me know if there is anything I can do to help -- would love to use this!

l1nxy commented:
> @l1nxy how is this going? Let me know if there is anything I can do to help -- would love to use this!

Apologies for the delayed update. I have been busy and unable to dedicate time to this. If you would like to proceed, feel free to do so, and I will close the pr.

l1nxy commented:
@s-i-l-k-e Hi, I have time now and will continue working on this.

l1nxy commented:
@tustvold   Hi, sorry for the delayed update, i have a question and hope you can help me. my question is which part modification will cause a breaking change, can i add GcpSigningCredential provider to GoogleCloudStorageBuilder and GoogleStorageConfig?

tustvold commented:
So long as you don't change an existing public API it won't be a breaking change. So you could add additional fields or new methods, but not change existing ones

l1nxy commented:
@tustvold Hi, I tested it using my GCP account. The ServiceAccount approach yielded correct results, but testing the email from the meta server was not done. Perhaps adding more unit tests or addressing some code issues. maybe i can add sign by HMAC key.

the test result is: 
![Snipaste_2024-04-01_18-46-52](https://github.com/apache/arrow-rs/assets/12152770/133f0ada-022c-4daa-a664-0b5204fab5a4)

tustvold reviewed (commented):
This is looking really nice, I took the liberty of pushing some relatively minor API tweaks. I'll test this out tomorrow, and assuming that all works get it merged

tustvold commented on object_store/src/gcp/credential.rs:
```diff
@@ -689,10 +692,12 @@ impl GCSAuthorizer {
-            Some(key) => key.sign(&string_to_sign)?,
-            None => client.sign_blob(&string_to_sign, email).await?,
-        };
+        // let signature = match &self.credential.private_key {
```
?

tustvold commented on object_store/src/gcp/credential.rs:
```diff
@@ -44,7 +44,10 @@ use std::time::{Duration, Instant};
-pub const DEFAULT_SCOPE: &str = "https://www.googleapis.com/auth/devstorage.full_control";
+pub const DEFAULT_SCOPE: [&str; 2] = [
+    "https://www.googleapis.com/auth/devstorage.full_control",
+    "https://www.googleapis.com/auth/cloud-platform",
```
What is the motivation for this change?

tustvold commented:
I've tested this and it appears to be working correctly :tada: 

Thank you for this, epic work

alamb commented:
Thank you @l1nxy  and @tustvold  -- this is a very nice feature
